src/D2L.CodeStyle.Analyzers/Language/ParamNamesShouldMatchInterfaceAnalyzer.cs

```diff
@@ -0,0 +1,156 @@
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalyzer );
+		}
+
+		private static void RegisterAnalyzer( CompilationStartAnalysisContext context ) {
+			context.RegisterSymbolAction(
+				ctx => AnalyzeClassDefinition(
+					ctx,
+					(INamedTypeSymbol)ctx.Symbol
+				),
+				SymbolKind.NamedType
+			);
+		}
+
+		private static void AnalyzeClassDefinition(
+			SymbolAnalysisContext context,
+			INamedTypeSymbol typeSymbol
+		) {
+			if( !typeSymbol.IsDefinition ) {
+				return;
+			}
+
+			if( !typeSymbol.Interfaces.Any() ) {
+				return;
+			}
+
+			ImmutableArray<IMethodSymbol> interfaceMethods = CollectInterfaceMethods( typeSymbol );
+			ImmutableArray<IMethodSymbol> classMethods = CollectClassMethods( typeSymbol );
+			foreach( var interfaceMethod in interfaceMethods ) {
+				IMethodSymbol classMethod = MatchMethod( classMethods, interfaceMethod );
+
+				if( classMethod.Parameters.Length == 0 ) {
+					continue;
+				}
+
+				ImmutableArray<IParameterSymbol> classMethodParameters = classMethod.Parameters;
+				ImmutableArray<IParameterSymbol> interfaceMethodParameters = interfaceMethod.Parameters;
+				for( int i = 0; i < classMethodParameters.Length; ++i ) {
+					string classMethodParameterName = classMethodParameters[ i ].Name;
+					string interfaceMethodParameterName = interfaceMethodParameters[ i ].Name;
+
+					if( classMethodParameterName.Equals( interfaceMethodParameterName, StringComparison.InvariantCultureIgnoreCase ) ) {
+						continue;
+					}
+
+					Location location = classMethodParameters[ i ].DeclaringSyntaxReferences.First().GetSyntax().GetLocation();
+					context.ReportDiagnostic( Diagnostic.Create( Diagnostics.InterfaceImplementationParamNameMismatch, location, interfaceMethodParameterName ) );
+				}
+			}
+		}
+
+		private static ImmutableArray<IMethodSymbol> CollectInterfaceMethods(
+			INamedTypeSymbol typeSymbol
+		) {
+			var builder = ImmutableArray.CreateBuilder<IMethodSymbol>();
+
+			foreach( INamedTypeSymbol @interface in typeSymbol.AllInterfaces ) {
+				ImmutableArray<ISymbol> members = @interface.GetMembers();
+				ImmutableArray<IMethodSymbol> methods = members
+					.Where( s => s.Kind == SymbolKind.Method )
+					.Cast<IMethodSymbol>()
+					.Where( s => s.MethodKind == MethodKind.Ordinary )
+					.ToImmutableArray();
+
+				foreach( var method in methods ) {
+					builder.Add( method );
+				}
+			}
+
+			return builder.ToImmutable();
+		}
+
+		private static ImmutableArray<IMethodSymbol> CollectClassMethods(
+			INamedTypeSymbol typeSymbol
+		) {
+			return typeSymbol
+				.GetMembers()
+				.Where( s => s.Kind == SymbolKind.Method )
+				.Cast<IMethodSymbol>()
+				.Where( s =>
+					s.MethodKind == MethodKind.ExplicitInterfaceImplementation
+					|| (
+						s.MethodKind == MethodKind.Ordinary
+						&& s.DeclaredAccessibility == Accessibility.Public
+					)
+				)
+				.ToImmutableArray();
+		}
+
+		private static IMethodSymbol MatchMethod(

```

Comment:
Maybe the SemanticModel has stuff to help with this, e.g. getting the declaration for that symbol?